### ISSUE TITLE
kubernetes + acme open traefik/acme/account: not a directory

### ISSUE THREAD

**Issue by Thermi** [containous/traefik#2521]
### Do you want to request a *feature* or report a *bug*?

bug

### What did you do?

The following `traefik.toml` file is supposed to use the KV store (etcd here) to store the acme related stuff. But it doesn't do that. It instead tries to open this path in the file system, which obviously doesn't work.

```shell
time="2017-12-04T16:41:34Z" level=error msg="Error creating TLS config: open traefik/acme/account: not a directory" 
time="2017-12-04T16:41:34Z" level=fatal msg="Error preparing server: open traefik/acme/account: not a directory" 
```

### What did you expect to see?

That traefik uses the KV store.

### What did you see instead?

That it doesn't use it, breaking lets encrypt support.

### Output of `traefik version`: (_What version of Traefik are you using?_)

1.4.2

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
apiVersion: v1
data:
  traefik.toml: |
    defaultEntryPoints = ["http", "https"]

    # get config from kubernetes ingress configs
    [kubernetes]

    # tell traefik where to get the TLS cert for api.rimoto.net
    [entryPoints]
      [entryPoints.https]
      address = ":443"
        [entryPoints.https.tls]
            [[entryPoints.https.tls.certificates]]
            CertFile = "certificate"
            KeyFile = "key"
      [entryPoints.http]
        address = ":80"
    [web]
    address = ":8080"
    ReadOnly = true
    
    [acme]
    email = "myemail"
    storage = "traefik/acme/account"
    entryPoint = "https"
    onHostRule = true
    acmeLogging = true

    [[acme.domains]]
   <internal stuff 1>
    [[acme.domains]]
   <internal stuff 2>

kind: ConfigMap
metadata:
  name: traefik-config
```

Logs:

```shell
time="2017-12-04T16:43:23Z" level=info msg="Using TOML configuration file /etc/traefik/traefik.toml" 
legolog: 2017/12/04 16:44:44 [INFO] acme: Registering account for myemail
time="2017-12-04T16:44:45Z" level=error msg="Error creating TLS config: open traefik/acme/account: not a directory" 
time="2017-12-04T16:44:45Z" level=fatal msg="Error preparing server: open traefik/acme/account: not a directory" 
```

**Comment by nmengin** [comment]
Hello @Thermi . Many thanks for you interest in the project.

How I understand your Træfik configuration, the ETCD configuration is missing into you `traefik.toml` file.
That's why Træfik generates this error message.

You can add this kind of configuration : 
```
[etcd]
endpoint = "X.X.X.X:2379"
watch = true
prefix = "/traefik"
```
*Note :* if you migrate to the new release-candidate version (1.5-rc2), you will be able to use the ETCD V3 API by adding `useAPIV3 = true` to your ETCD configuration.

I close de issue but feel free to reopen it if it does not solve your problem.

**Comment by Thermi** [comment]
That's sad. Accessing etcd is not supposed to be possible by pods, so if we were to use it, we would need to enable traefik to use authentication for connecting to etcd and also somehow configure etcd, which is not always possible (quite probably not with GKE). So we're stuck. Is there any other way to use the let's encrypt support with  Kubernetes? etcd is the only way for traefik to share the information. Maybe there is some native way on Kubernetes by using its API for sharing the certificates? Maybe by using specific secrets? I have not found any way to do this on Kubernetes yet. Am I missing something?

**Comment by nmengin** [comment]
@Thermi WDYT about creating a pod which contains you ETCD (or other KV store) and allows Træfik to access to it?

**Comment by Thermi** [comment]
@nmengin The etcd server of Kubernetes runs on the (or all) the master nodes. It is not a pod. The list of pods and the complete configuration of the cluster is stored in the etcd server. So one can't do what you propose. AFAIK by Kubernetes design, only the kubelet processes that each node runs are supposed to ever connect to the etcd server.

**Comment by timoreimann** [comment]
@Thermi how about rolling your own etcd cluster, possibly inside Kubernetes? AFAIK there's a Helm chart and the etcd operator available to ease the management.

**Comment by Thermi** [comment]
@timoreimann I'd consider that the last resort. traefik already has the ability to do Kubernetes API calls, so why not use a configmap or secret for storing the secrets, too?

**Comment by timoreimann** [comment]
@Thermi I'm not too familiar with ACME. Is there a need to mutate any data that come with it? If no, then ConfigMaps or Secrets might be a viable option. Otherwise, we're probably better off with a real data storage solution.

Either way, Traefik currently does not support ACME from anything but a K/V store (like etcd) on Kubernetes. (@dtomcej, correct me if I I'm wrong here please.) This issue was about etcd in the first place, hence my suggestion to go with a dedicated cluster.

I totally don't mind considering a solution more tightly integrated with Kubernetes, but it should come as a separate feature request in my opinion.

**Comment by Thermi** [comment]
[The page about Let's Encrypt configuration states this](https://docs.traefik.io/configuration/acme/)
```
# File or key used for certificates storage.
#
# Required
#
storage = "acme.json"
# or `storage = "traefik/acme/account"` if using KV store.

```

The "storage" implies that it is actively written to, which kind of makes sense. In a clustered environment, the issued certificates need to be stored permanently and distributed somehow.


AFAIK kubernetes allows notifications to processes when a CM is changed. A secret likely, too. I guess it is a normal inotify mechanism caused by the simple replacement of the file. One could use that to get data from the other traefik processes. The only issue now is an atomic writing to the file via the Kubernetes API.

Well, I'll make a feature request as soon as we figure out what we can actually do to solve that.

**Comment by timoreimann** [comment]
@Thermi a ConfigMap/Secret is an object living on the Kubernetes API server, the file system is not involved. The notification mechanism is an event sent through to Traefik's Kubernetes client. The Kubernetes client library already comes with a leadership implementation which could serve as a way to do atomic writes.

There are likely a few other challenges to it, but it seems doable. Feel free to go ahead and file a feature request.

**Comment by Thermi** [comment]
@timoreimann [That is wrong.](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically)

**Comment by timoreimann** [comment]
@Thermi nice, I didn't know that. (Is that a newer feature by any chance?)

I wonder if we'd be able to benefit from this for free, or whether we need to make sure our version of client-go is up-to-date.

**Comment by Thermi** [comment]
@timoreimann No, it's not new. It was like that since at least 1.4.

You can't write to configmaps, because they're read only, so at least some code for the writing is needed.
I suppose just updating the lib before the next build will be enough?

**Comment by timoreimann** [comment]
@Thermi yeah we can probably do that. Let's continue discussing in that to-be-created issue and make sure we keep track of the known/presumed requirements.

**Comment by nmengin** [comment]
Closed in favor of #2542
